src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -501,5 +573,11 @@ internal async ValueTask<StructuredAnalyzerConfigOptions> GetAnalyzerConfigOptio
             var provider = (ProjectState.ProjectAnalyzerConfigOptionsProvider)Project.State.AnalyzerOptions.AnalyzerConfigOptionsProvider;
             return await provider.GetOptionsAsync(DocumentState, cancellationToken).ConfigureAwait(false);
         }
+
+        private static WeakReference<SemanticModel>? SetModel(ref WeakReference<SemanticModel>? reference, SemanticModel result)
+        {
+            var original = Interlocked.CompareExchange(ref reference, new WeakReference<SemanticModel>(result), null);
+            return original;
```

Comment:
this method can be inlined to the call site, since you only need to call it once.